feat(NODE-4738)!: remove dot notation support by default (#3520)
mongodb/node-mongodb-native: etc/notes/CHANGES_5.0.0.md

## Changes

### Dot Notation Typescript Support Removed By Default

> **NOTE** This version only removes our **Typescript** compile-time support for dot notation in queries.

Reviewer: ```suggestion
**NOTE** This is a **Typescript compile-time only** change. Dot notation in filters sent to MongoDB still work the same. 
```

> **NOTE** This is a **Typescript compile-time only** change. Dot notation in filters sent to MongoDB still work the same. 

Reviewer 2: ```suggestion
**NOTE** This is a **Typescript compile-time only** change. Dot notation in filters sent to MongoDB will still work the same. 
```

Version 4.3.0 introduced Typescript support for dot notation in filter predicates.  For example:

```typescript
interface Schema {
  user: {
    name: string
  }
}

declare const collection: Collection<Schema>;
// compiles pre-v4.3.0, fails in v4.3.0+
collection.find({ 'user.name': 4 })
```

> This change caused many problems for users, including slow compilation times and compile errors for

Reviewer 2: ```suggestion
This change caused a number of problems for users, including slow compilation times and compile errors for
```

> valid dot notation queries.

Reviewer 2: ```suggestion
valid dot notation queries. While we have tried to mitigate this issue as much as possible
in v4, ultimately we do not believe that this feature is fully production ready for all use cases.
```

> Driver 5.0 removes type checking on filter predicates.  The preceding example compile will compile with

Reviewer 3: ```suggestion
Driver 5.0 removes type checking on filter predicates.  The preceding example will compile with
```

> Driver 5.0 removes type checking on filter predicates.  The preceding example will compile with

Reviewer 2: Not all type checking, right, just dot notation type checking?

Author: Yeah, just dot notation support

driver v5.

#### Dot Notation Helper Types Exported

> Although we removed support for type checking on dot notation filters by default, we now export these types.

Reviewer 2: ```suggestion
Although we removed support for type checking on dot notation filters by default, we have preserved the
corresponding types in an experimental capacity.
```

> These helper types can be used for type checking.  We export the `DotNotationUpdateFilter` and the `DotNotationFilter`

Reviewer: ```suggestion
These helper types can be used for type checking.  We export the `StrictUpdateFilter` and the `StrictFilter`
```

types for type safety in updates and finds.

> To use one of the new types, simply create a predicate that uses dot notation and assign it the type of `DotNotationFilter<your schema>`.

Reviewer: ```suggestion
To use one of the new types, simply create a predicate that uses dot notation and assign it the type of `StrictFilter<your schema>`.
```

```typescript
interface Schema {
  user: {
    name: string
  }
}

declare const collection: Collection<Schema>;

// fails to compile, 4 is not assignable to type "string"

> const filterPredicate: DotNotationFilter<Schema> = { 'user.name': 4 };

Reviewer: ```suggestion
const filterPredicate: StrictFilter<Schema> = { 'user.name': 4 };
```

collection.find(filterPredicate);
```

> **NOTE** These types are also now marked experimental and can be changed at any time.

Reviewer 2: ```suggestion
**NOTE** As an experimental feature, these types can change at any time and are not recommended for production settings.
```

## Changes

### Dot Notation Typescript Support Removed By Default

**NOTE** This is a **Typescript compile-time only** change. Dot notation in filters sent to MongoDB will still work the same. 

Version 4.3.0 introduced Typescript support for dot notation in filter predicates.  For example:

```typescript
interface Schema {
  user: {
    name: string
  }
}

declare const collection: Collection<Schema>;
// compiles pre-v4.3.0, fails in v4.3.0+
collection.find({ 'user.name': 4 })
```

This change caused a number of problems for users, including slow compilation times and compile errors for
valid dot notation queries. While we have tried to mitigate this issue as much as possible
in v4, ultimately we do not believe that this feature is fully production ready for all use cases.

> Driver 5.0 removes type checking on filter predicates.  The preceding example will compile with

Author: ```suggestion
Driver 5.0 removes type checking for dot notation on filter predicates.  The preceding example will compile with
```

> Driver 5.0 removes type checking for dot notation on filter predicates.  The preceding example will compile with

Author: ```suggestion
Driver 5.0 removes type checking for dot notation in filter predicates.  The preceding example will compile with
```
